### PR TITLE
added relion version 3 (beta)

### PR DESCRIPTION
**Pull request by JusticeForMikeBrown**
[Please see ](https://bitbucket.org/scheres/relion-3.0_beta/overview)

### CODE REVIEW

File: var/spack/repos/builtin/packages/relion/package.py [hunk 1]
```diff
@@ -52,14 +53,18 @@ class Relion(CMakePackage, CudaPackage):
     depends_on('fltk', when='+gui')
     # cuda 9 not yet supported
     #  https://github.com/3dem/relion/issues/296
-    depends_on('cuda@8.0:8.99', when='+cuda')
+    # depends_on('cuda@8.0:8.99', when='+cuda')
+    depends_on('cuda@9:', when='version==beta-3')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/relion/package.py]: So version `beta-3` requires `cuda` regardless of whether or not the `+cuda` variant is enabled and no other version requires `cuda` even when `+cuda`?

File: var/spack/repos/builtin/packages/relion/package.py [hunk 2]
```diff
@@ -32,11 +32,12 @@ class Relion(CMakePackage, CudaPackage):
     electron cryo-microscopy (cryo-EM)."""
 
     homepage = "http://http://www2.mrc-lmb.cam.ac.uk/relion"
-    git      = "https://github.com/3dem/relion.git"
+    url      = "https://github.com/3dem/relion"
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/relion/package.py]: Please undo this change.

File: var/spack/repos/builtin/packages/relion/package.py [hunk 3]
```diff
@@ -32,11 +32,12 @@ class Relion(CMakePackage, CudaPackage):
...
 
-    version('develop')
-    version('2.1', tag='2.1')
-    version('2.0.3', tag='2.0.3')
+    version('2.1', git='https://github.com/3dem/relion.git', tag='2.1')
+    version('2.0.3', git='https://github.com/3dem/relion.git', tag='2.0.3')
+    version('develop', git='https://github.com/3dem/relion.git')
+    version('beta-3', git='https://bitbucket.org/scheres/relion-3.0_beta.git')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/relion/package.py]: Why not call the version `3.0_beta` and mark `2.1` as `preferred=True`?

File: var/spack/repos/builtin/packages/relion/package.py [hunk 4]
```diff
@@ -52,14 +53,18 @@ class Relion(CMakePackage, CudaPackage):
...
     # use gcc < 5 when compiled with cuda 8
     conflicts('%gcc@5:', when='+cuda')
 
     def cmake_args(self):
         args = [
             '-DCMAKE_C_FLAGS=-g',
             '-DCMAKE_CXX_FLAGS=-g',
+            '-DALTCPU=ON',
+            '-DFORCE_OWN_TBB=ON',
+            '-DMKLFFT=ON',
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/relion/package.py]: Just to make sure, these flags all apply to the older version as well?

File: var/spack/repos/builtin/packages/relion/package.py (hunk 3 above)
Comment (JusticeForMikeBrown) [reply var/spack/repos/builtin/packages/relion/package.py]: Been a while since added things to Spack... Thanks!

File: var/spack/repos/builtin/packages/relion/package.py (hunk 2 above)
Comment (JusticeForMikeBrown) [reply var/spack/repos/builtin/packages/relion/package.py]: My mistake!

File: var/spack/repos/builtin/packages/relion/package.py (hunk 4 above)
Comment (JusticeForMikeBrown) [reply var/spack/repos/builtin/packages/relion/package.py]: Good question...

File: var/spack/repos/builtin/packages/relion/package.py [hunk 5]
```diff
@@ -32,11 +32,12 @@ class Relion(CMakePackage, CudaPackage):
     electron cryo-microscopy (cryo-EM)."""
 
     homepage = "http://http://www2.mrc-lmb.cam.ac.uk/relion"
-    git      = "https://github.com/3dem/relion.git"
+    url      = "https://github.com/3dem/relion.git"
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/relion/package.py]: This change still needs to be reverted.

File: var/spack/repos/builtin/packages/relion/package.py [hunk 6]
```diff
@@ -32,11 +32,12 @@ class Relion(CMakePackage, CudaPackage):
...
 
-    version('develop')
-    version('2.1', tag='2.1')
-    version('2.0.3', tag='2.0.3')
+    version('2.1', git='https://github.com/3dem/relion.git', preferred='true', tag='2.1')
+    version('2.0.3', git='https://github.com/3dem/relion.git', tag='2.0.3')
+    version('develop', git='https://github.com/3dem/relion.git')
+    version('3.0_beta', git='https://bitbucket.org/scheres/relion-3.0_beta.git')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/relion/package.py]: Can you put this above 2.1?

File: var/spack/repos/builtin/packages/relion/package.py [hunk 7]
```diff
@@ -50,9 +51,13 @@ class Relion(CMakePackage, CudaPackage):
...
     depends_on('fltk', when='+gui')
-    # cuda 9 not yet supported
-    #  https://github.com/3dem/relion/issues/296
-    depends_on('cuda@8.0:8.99', when='+cuda')
+
+    # relion 3 supports cuda 9
+    if ('@:3.0_beta'):
+        depends_on('cuda@9:', when='+cuda')
+    else:
+        depends_on('cuda@8.0:8.99', when='+cuda')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/relion/package.py]: This won't work, non-empty strings will always evaluate to True in Python. Try:
```python
depends_on('cuda', when='+cuda')
depends_on('cuda@9:', when='@3: +cuda')
depends_on('cuda@8.0:8.99', when='@:2 +cuda')
```

File: var/spack/repos/builtin/packages/relion/package.py [hunk 8]
```diff
@@ -70,9 +75,17 @@ def cmake_args(self):
...
             args += [
                 '-DCUDA=on',
+                '-DCudaTexture=ON',
             ]
             if carch is not None:
                 args += [
                     '-DCUDA_ARCH=%s' % (carch),
                 ]
+
+	if ('@:3.0_beta'):
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/relion/package.py]:
```python
if self.spec.satisfies('@3:'):
```

**Comment by JusticeForMikeBrown** [comment]
Hi @adamjstewart 

I have made all the changes you request though this fails:

`spack install relion@3.0_beta+cuda cuda_arch=60 ^openmpi+cuda %gcc@6.3.0`

[Per issue I opened](https://github.com/3dem/relion/issues/380).

I have tried multiple versions of cmake, gcc, openmpi, but the issue persists during Spack install phase.

The issue seems to be with pthreads which was found during `configure`:

```
-- Setting accelerated code precision to single
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - not found
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
```

CMake error log contains:


```
Determining if the pthread_create exist failed with the following output:
Change Dir: /tmp/user/spack-stage/spack-stage-i6mWLW/relion-3.0_beta/spack-build/CMakeFiles/CMakeTmp

Run Build Command:"/usr/bin/gmake" "cmTC_23a1c/fast"
/usr/bin/gmake -f CMakeFiles/cmTC_23a1c.dir/build.make CMakeFiles/cmTC_23a1c.dir/build
gmake[1]: Entering directory `/tmp/user/spack-stage/spack-stage-i6mWLW/relion-3.0_beta/spack-build/CMakeFiles/CMakeTmp'
Building C object CMakeFiles/cmTC_23a1c.dir/CheckSymbolExists.c.o
/pbtech_mounts/homes027/user/prepos/spack/lib/spack/env/gcc/gcc   -fPIC    -o CMakeFiles/cmTC_23a1c.dir/CheckSymbolExists.c.o   -c /tmp/user/spack-stage/spack-stage-i6mWLW/relion-3.0_beta/spack-build/CMakeFiles/CMakeTmp/CheckSymbolExists.c
Linking C executable cmTC_23a1c
/pbtech_mounts/homes027/user/prepos/spack/opt/spack/linux-centos7-x86_64/gcc-6.3.0/cmake-3.12.0-d6i3ujxpf5m4aiggveyzridh7rxbb6bh/bin/cmake -E cmake_link_script CMakeFiles/cmTC_23a1c.dir/link.txt --verbose=1
/pbtech_mounts/homes027/user/prepos/spack/lib/spack/env/gcc/gcc -fPIC     -rdynamic CMakeFiles/cmTC_23a1c.dir/CheckSymbolExists.c.o  -o cmTC_23a1c
CMakeFiles/cmTC_23a1c.dir/CheckSymbolExists.c.o: In function `main':
**CheckSymbolExists.c:(.text+0x1b): undefined reference to `pthread_create'
collect2: error: ld returned 1 exit status**
gmake[1]: *** [cmTC_23a1c] Error 1
gmake[1]: Leaving directory `/tmp/user/spack-stage/spack-stage-i6mWLW/relion-3.0_beta/spack-build/CMakeFiles/CMakeTmp'
gmake: *** [cmTC_23a1c/fast] Error 2

File /tmp/user/spack-stage/spack-stage-i6mWLW/relion-3.0_beta/spack-build/CMakeFiles/CMakeTmp/CheckSymbolExists.c:
/* */
#include <pthread.h>

int main(int argc, char** argv)
{
  (void)argv; 
#ifndef pthread_create
  return ((int*)(&pthread_create))[argc];
#else
  (void)argc;
  return 0;
#endif
}

Determining if the function pthread_create exists in the pthreads failed with the following output:
Change Dir: /tmp/user/spack-stage/spack-stage-i6mWLW/relion-3.0_beta/spack-build/CMakeFiles/CMakeTmp

Run Build Command:"/usr/bin/gmake" "cmTC_16571/fast"
/usr/bin/gmake -f CMakeFiles/cmTC_16571.dir/build.make CMakeFiles/cmTC_16571.dir/build
gmake[1]: Entering directory `/tmp/user/spack-stage/spack-stage-i6mWLW/relion-3.0_beta/spack-build/CMakeFiles/CMakeTmp'
Building C object CMakeFiles/cmTC_16571.dir/CheckFunctionExists.c.o
/pbtech_mounts/homes027/user/prepos/spack/lib/spack/env/gcc/gcc   -fPIC -DCHECK_FUNCTION_EXISTS=pthread_create   -o CMakeFiles/cmTC_16571.dir/CheckFunctionExists.c.o   -c /pbtech_mounts/homes027/user/prepos/spack/opt/spack/linux-centos7-x86_64/gcc-6.3.0/cmake-3.12.0-d6i3ujxpf5m4aiggveyzridh7rxbb6bh/share/cmake-3.12/Modules/CheckFunctionExists.c
Linking C executable cmTC_16571
/pbtech_mounts/homes027/user/prepos/spack/opt/spack/linux-centos7-x86_64/gcc-6.3.0/cmake-3.12.0-d6i3ujxpf5m4aiggveyzridh7rxbb6bh/bin/cmake -E cmake_link_script CMakeFiles/cmTC_16571.dir/link.txt --verbose=1
/pbtech_mounts/homes027/user/prepos/spack/lib/spack/env/gcc/gcc -fPIC -DCHECK_FUNCTION_EXISTS=pthread_create    -rdynamic CMakeFiles/cmTC_16571.dir/CheckFunctionExists.c.o  -o cmTC_16571 -lpthreads
**/usr/bin/ld: cannot find -lpthreads**
collect2: error: ld returned 1 exit status
gmake[1]: *** [cmTC_16571] Error 1
gmake[1]: Leaving directory `/tmp/user/spack-stage/spack-stage-i6mWLW/relion-3.0_beta/spack-build/CMakeFiles/CMakeTmp'
gmake: *** [cmTC_16571/fast] Error 2
```

Any idea on what could be causing this problem?  I greatly appreciate your continual help.

**Comment by JusticeForMikeBrown** [comment]
OK, I was able to compile with lower cmake@3.9.4

File: var/spack/repos/builtin/packages/relion/package.py [hunk 9]
```diff
@@ -65,14 +92,20 @@ def cmake_args(self):
...
-            args += [
-                '-DCUDA=on',
-            ]
-            if carch is not None:
-                args += [
-                    '-DCUDA_ARCH=%s' % (carch),
-                ]
+            # relion+cuda requires selecting cuda_arch
+	    if not carch: 
+		sys.stdout.write('relion+cuda requires selecting cuda_arch')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/relion/package.py]: This should be a `conflicts` directive.

File: var/spack/repos/builtin/packages/relion/package.py [hunk 10]
```diff
@@ -65,14 +92,20 @@ def cmake_args(self):
...
+		sys.stdout.flush()
+	        sys.exit()
+	    else:
+                args += ['-DCUDA=ON','-DCudaTexture=ON', '-DCUDA_ARCH=%s' % (carch)]
+		sys.stdout.write('cuda arch ' + carch + ' selected')
+		sys.stdout.flush()
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/relion/package.py]: I don't think we need to tell the user what they selected. Also, `print` is much simpler and does the same thing.

File: var/spack/repos/builtin/packages/relion/package.py [hunk 11]
```diff
@@ -34,29 +35,55 @@ class Relion(CMakePackage, CudaPackage):
     homepage = "http://http://www2.mrc-lmb.cam.ac.uk/relion"
     git      = "https://github.com/3dem/relion.git"
 
-    version('develop')
-    version('2.1', tag='2.1')
-    version('2.0.3', tag='2.0.3')
+    version('3.0_beta', git='https://bitbucket.org/scheres/relion-3.0_beta.git')
+    version('2.1', git='https://github.com/3dem/relion.git', preferred='true', tag='2.1')
+    version('2.0.3', git='https://github.com/3dem/relion.git', tag='2.0.3')
+    version('develop', git='https://github.com/3dem/relion.git')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/relion/package.py]: These version-specific git URLs can be removed unless they are different than the package-level git.

File: var/spack/repos/builtin/packages/relion/package.py [hunk 12]
```diff
@@ -34,29 +35,55 @@ class Relion(CMakePackage, CudaPackage):
...
     variant('gui', default=True, description="build the gui")
     variant('cuda', default=True, description="enable compute on gpu")
     variant('double', default=True, description="double precision (cpu) code")
     variant('double-gpu', default=False, description="double precision (gpu) code")
+    # added below cluster and desktop variants given some shared libraries in /usr/lib, like gpfs,
+    # are loaded by ld.so after compile which makes relion not work on desktops which may
+    # not be a member of gpfs cluster thus lack these given libraries
+    # https://github.com/spack/spack/issues/8384
+    variant('cluster', default=True, description="build relion for use in cluster")
+    variant('desktop', default=False, description="build relion for use in desktop")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/relion/package.py]: Are these mutually exclusive?

File: var/spack/repos/builtin/packages/relion/package.py [hunk 13]
```diff
@@ -34,29 +35,55 @@ class Relion(CMakePackage, CudaPackage):
...
+    # relion 3 supports cuda 9
+    # relion < 3 does not
+    depends_on('cuda', when='+cuda')
+    depends_on('cuda@9:', when='@3: +cuda')
+    depends_on('cuda@8.0:8.99', when='@:2 +cuda')
+
+    # use gcc 4 when using cuda8 
+    # use up to gcc 6 when using cuda9
+    conflicts('%gcc@7:', when='@3: +cuda')
+    conflicts('%gcc@5:', when='@:2 +cuda')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/relion/package.py]: Is this a restriction on building relion or a restriction on building cuda?

File: var/spack/repos/builtin/packages/relion/package.py [hunk 14]
```diff
@@ -32,36 +33,57 @@ class Relion(CMakePackage, CudaPackage):
...
 
     def cmake_args(self):
+        
+        # this creates string such as ('60',)
+	# which will never evaluate properly
+        c = str(self.spec.variants['cuda_arch'].value)
+
+        # find value between single quotes
+	carch = c[c.find("'")+1:c.rfind("'")]
```
[the remark this answers is not in the document]
Comment (JusticeForMikeBrown) [reply var/spack/repos/builtin/packages/relion/package.py]: It's not a list.

File: var/spack/repos/builtin/packages/relion/package.py (hunk 10 above)
Comment (JusticeForMikeBrown) [reply var/spack/repos/builtin/packages/relion/package.py]: I tried print but it only prints to the spack build log which does not appear when I then do sys.exit()

File: var/spack/repos/builtin/packages/relion/package.py (hunk 12 above)
Comment (JusticeForMikeBrown) [reply var/spack/repos/builtin/packages/relion/package.py]: Yes

File: var/spack/repos/builtin/packages/relion/package.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/relion/package.py]: What type is it? It sure looks like a tuple...

File: var/spack/repos/builtin/packages/relion/package.py (hunk 12 above)
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/relion/package.py]: Then this should be a single multi-valued variant.

File: var/spack/repos/builtin/packages/relion/package.py (hunk 11 above)
Comment (JusticeForMikeBrown) [reply var/spack/repos/builtin/packages/relion/package.py]: Fixing now!

File: var/spack/repos/builtin/packages/relion/package.py (hunk 12 above)
Comment (JusticeForMikeBrown) [reply var/spack/repos/builtin/packages/relion/package.py]: Adding conflicts

File: var/spack/repos/builtin/packages/relion/package.py (hunk 9 above)
Comment (JusticeForMikeBrown) [reply var/spack/repos/builtin/packages/relion/package.py]: Sorry, @adamjstewart I have no idea how to do this

Can conflicts() go under the cmake_args function?  

I tried:
```
    if '+cuda' in self.spec:
            # relion+cuda requires selecting cuda_arch
            if not carch:
                conflicts('%gcc')
```
For example which does not stop execution.

File: var/spack/repos/builtin/packages/relion/package.py (hunk 13 above)
Comment (JusticeForMikeBrown) [reply var/spack/repos/builtin/packages/relion/package.py]: It's a cuda restriction but it was easier for me to put the conflicts in this spec

File: var/spack/repos/builtin/packages/relion/package.py (hunk 9 above)
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/relion/package.py]: Nope, it has to be a top-level directive.

File: var/spack/repos/builtin/packages/relion/package.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (JusticeForMikeBrown) [reply var/spack/repos/builtin/packages/relion/package.py]: isinstance says it's a tuple!

[the remark this answers is not in the document]
Comment (JusticeForMikeBrown) [reply var/spack/repos/builtin/packages/relion/package.py]: So I apparently need to do `spack clean --all` after doing `spack edit relion` as the changes to these vars were not being reflected.  Hence why I didn't see this before.

File: var/spack/repos/builtin/packages/relion/package.py (hunk 13 above)
Comment (JusticeForMikeBrown) [reply var/spack/repos/builtin/packages/relion/package.py]: Hi @adamjstewart 

Can these conflicts remain?

File: var/spack/repos/builtin/packages/relion/package.py (hunk 9 above)
Comment (JusticeForMikeBrown) [reply var/spack/repos/builtin/packages/relion/package.py]: How do I reference spec.variants['cuda_arch'].value?
```
      test = spec.variants['cuda_arch'].value
      print(test)
      sys.exit()
  ```

  File "/home/user/prepos/spack/var/spack/repos/builtin/packages/relion/package.py", line 30, in <module>
    """RELION (for REgularised LIkelihood OptimisatioN, pronounce rely-on) is a
  File "/home/dod2014/prepos/spack/var/spack/repos/builtin/packages/relion/package.py", line 81, in Relion
    print(test)
NameError: name 'spec' is not defined

Comment (JusticeForMikeBrown) [reply var/spack/repos/builtin/packages/relion/package.py]: Think I have an idea.  Always `grep` before asking questions.

Comment (JusticeForMikeBrown) [reply var/spack/repos/builtin/packages/relion/package.py]: Never mind.

How do I use `spec.variants` outside the cmake function?

Comment (JusticeForMikeBrown) [reply var/spack/repos/builtin/packages/relion/package.py]: `spack.variant.spack.spec` 

I need to start using Eclipse or make vim a little more useful.

File: var/spack/repos/builtin/packages/relion/package.py (hunk 13 above)
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/relion/package.py]: Shouldn't they be added to the Cuda package instead of every package that depends on Cuda?

**Comment by JusticeForMikeBrown** [comment]
Hi

Can we now merge?

File: var/spack/repos/builtin/packages/relion/package.py [hunk 15]
```diff
@@ -34,29 +34,46 @@ class Relion(CMakePackage, CudaPackage):
...
     git      = "https://github.com/3dem/relion.git"
 
-    version('develop')
-    version('2.1', tag='2.1')
-    version('2.0.3', tag='2.0.3')
+    version('3.0_beta',
+            git='https://bitbucket.org/scheres/relion-3.0_beta.git')
+    version('2.1', git='https://github.com/3dem/relion.git', preferred='true')
+    version('2.0.3', git='https://github.com/3dem/relion.git')
+    version('develop', git='https://github.com/3dem/relion.git')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/relion/package.py]: Version 3.0_beta is the only one that needs to specify `git=`. The rest can use the out package-level git.

Comment (JusticeForMikeBrown) [reply var/spack/repos/builtin/packages/relion/package.py]: I am a bit confused since  `https://bitbucket.org/scheres/relion-3.0_beta.git` does not contain code for previous versions.

Version 2 and lower are at `https://github.com/3dem/relion.git`

Comment (JusticeForMikeBrown) [reply var/spack/repos/builtin/packages/relion/package.py]: Oh, I see!

`git = https://github.com/3dem/relion.git` has previous versions.

File: var/spack/repos/builtin/packages/relion/package.py [hunk 16]
```diff
@@ -34,29 +34,46 @@ class Relion(CMakePackage, CudaPackage):
     homepage = "http://http://www2.mrc-lmb.cam.ac.uk/relion"
     git      = "https://github.com/3dem/relion.git"
 
+    version('3.0_beta',
+            git='https://bitbucket.org/scheres/relion-3.0_beta.git')
+    version('2.1', preferred='true')
+    version('2.0.3')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/relion/package.py]: Does this still work if you don't specify the tag? It isn't going to automatically infer where to download `2.0.3`.

File: var/spack/repos/builtin/packages/relion/package.py [hunk 17]
```diff
@@ -34,29 +34,46 @@ class Relion(CMakePackage, CudaPackage):
...
 
     variant('gui', default=True, description="build the gui")
     variant('cuda', default=True, description="enable compute on gpu")
     variant('double', default=True, description="double precision (cpu) code")
-    variant('double-gpu', default=False, description="double precision (gpu) code")
+    variant('double-gpu', default=False, description="double precision gpu")
+    # added below cluster and desktop variants given some shared libraries in
+    # /usr/lib, like gpfs, loaded by ld.so after compile which makes relion
+    # not work on desktops which may not be a member of gpfs cluster
+    # thus lack these given libraries
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/relion/package.py]: I feel like this is a run-on sentence...

Maybe something like:

> If built with `purpose=cluster`, automatically links to gpfs libraries, which may not be present on a desktop. For use on a desktop, select `purpose=desktop`.

File: var/spack/repos/builtin/packages/relion/package.py [hunk 18]
```diff
@@ -65,14 +82,19 @@ def cmake_args(self):
...
-                '-DCUDA=on',
-            ]
-            if carch is not None:
-                args += [
-                    '-DCUDA_ARCH=%s' % (carch),
-                ]
+            # relion+cuda requires selecting cuda_arch
+            if not carch:
+                # below does not work so raising an exception
+                conflicts('%gcc@4.0:6.99', msg='you must select cuda_arch')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/relion/package.py]: This won't work. `conflicts` can only be used at the class-level, it can't be used inside of a function.

There isn't a "not" in the spec syntax, and I don't know of a way to say `if cuda_arch=''`. You'll just have to raise a ValueError here with this message.

Comment (JusticeForMikeBrown) [reply var/spack/repos/builtin/packages/relion/package.py]: Thanks

If the user does not define cuda_arch=value at command line then `carch` isn't defined so then the `if not` does indeed work.

File: var/spack/repos/builtin/packages/relion/package.py (hunk 16 above)
Comment (JusticeForMikeBrown) [reply var/spack/repos/builtin/packages/relion/package.py]: Yes works for `2.0.3`

`Cloning git repository: https://github.com/3dem/relion.git`

Comment (JusticeForMikeBrown) [reply var/spack/repos/builtin/packages/relion/package.py]: Adding tag='2.0.3' back while also removing 'develop' version as it no longer exists.

File: var/spack/repos/builtin/packages/relion/package.py [hunk 19]
```diff
@@ -62,7 +62,6 @@ class Relion(CMakePackage, CudaPackage):
 
     # relion 3 supports cuda 9
     # relion < 3 does not
-    depends_on('cuda', when='+cuda')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/relion/package.py]: You should keep this line. Without it, it becomes more difficult to specify which CUDA version you want (you would have to explicitly state the `relion` version). This is a workaround for a bug in Spack's concretization mechanism.

File: var/spack/repos/builtin/packages/relion/package.py [hunk 20]
```diff
@@ -34,29 +34,42 @@ class Relion(CMakePackage, CudaPackage):
     homepage = "http://http://www2.mrc-lmb.cam.ac.uk/relion"
     git      = "https://github.com/3dem/relion.git"
 
-    version('develop')
-    version('2.1', tag='2.1')
+    version('3.0_beta',
+            git='https://bitbucket.org/scheres/relion-3.0_beta.git')
+    version('2.1', preferred='true')
```
[the remark this answers is not in the document]
Comment (JusticeForMikeBrown) [reply var/spack/repos/builtin/packages/relion/package.py]: They're back

File: var/spack/repos/builtin/packages/relion/package.py [hunk 21]
```diff
@@ -65,14 +78,18 @@ def cmake_args(self):
...
-                '-DCUDA=on',
-            ]
-            if carch is not None:
-                args += [
-                    '-DCUDA_ARCH=%s' % (carch),
-                ]
+            # relion+cuda requires selecting cuda_arch
+            if not carch:
+                print("you must select cuda_arch")
+                raise ValueError
```
[the remark this answers is not in the document]
Comment (JusticeForMikeBrown) [reply var/spack/repos/builtin/packages/relion/package.py]: Thanks!

File: var/spack/repos/builtin/packages/relion/package.py [hunk 22]
```diff
@@ -34,29 +34,42 @@ class Relion(CMakePackage, CudaPackage):
     homepage = "http://http://www2.mrc-lmb.cam.ac.uk/relion"
     git      = "https://github.com/3dem/relion.git"
 
-    version('develop')
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/relion/package.py]: I would keep the `develop` version, maybe be explicit where it's coming from:
```python
version('develop', branch='master')
```

Comment (JusticeForMikeBrown) [reply var/spack/repos/builtin/packages/relion/package.py]: Sounds good.  Relion's `master` branch should really be considered `develop` so nothing wrong with being more explicit.
